code/modules/mob/living/carbon/human/species/_species.dm

```diff
@@ -170,40 +171,109 @@
+
+	if(!regen_all && OS) //Only species-swap stuff that's ours and wasn't frankensteined or augmented.
+		var/list/remove_after = list("limbs" = list(), "organs" = list()) //Useful when going from species with more organs to species with less.
+		var/list/oldspecies_organs = OS.has_organ.Copy()
+		if(OS.mutantears)
+			oldspecies_organs["ears"] = OS.mutantears
+
+		/**HANDLE INTERNAL ORGANS**/
+		for(var/organ_type in species_organs) //Account for switching from a species with less organs to a species with more and vise versa.
+			if(!(organ_type in oldspecies_organs))
+				var/obj/item/organ/internal/I = H.get_organ_slot(organ_type)
+				if(!I)
+					replace_after["organs"][organ_type] = TRUE //Error 404 no type found. Will be generated from scratch later.
+				else //Liquidate the surplus.
+					remove_after["organs"] |= I
+					H.internal_organs -= I
+					H.internal_organs_slot -= I.slot
+					I.become_orphan()
+
+		for(var/O in H.internal_organs) //Same as above but for innards. Gotta go inside out to avoid runtimes.
+			var/obj/item/organ/internal/I = O
+			if(!(I.slot in oldspecies_organs) && (I.type != oldspecies_organs[I.slot]))
+				continue
+			if(I.is_robotic() && isrobot(initial(I))) //Were ya born that way?
+				continue
+			if((I?.dna.species.name != OS.name) || (I?.dna.real_name != H.real_name)) //Same slot, different species?
+				continue
+			if((I.slot in oldspecies_organs) && !(I.slot in species_organs))
+				remove_after["organs"] |= I
+			else
+				replace_after["organs"][I.slot] = I
 
-	LAZYREINITLIST(H.bodyparts)
-	LAZYREINITLIST(H.bodyparts_by_name)
-	LAZYREINITLIST(H.internal_organs)
+			H.internal_organs -= I
+			H.internal_organs_slot -= I.slot
+			I.become_orphan()
 
-	for(var/limb_type in has_limbs)
-		var/list/organ_data = has_limbs[limb_type]
-		var/limb_path = organ_data["path"]
-		var/obj/item/organ/O = new limb_path(H)
-		organ_data["descriptor"] = O.name
+		QDEL_LIST(remove_after["organs"]) //This needs to be broken out here because it made the loops above unable to read properties properly.
 
-	for(var/index in has_organ)
-		var/organ = has_organ[index]
-		// organ new code calls `insert` on its own
-		new organ(H)
+		/**HANDLE LIMBS**/
+		for(var/limb_type in has_limbs) //Just doing this 'cause I don't know what the future holds. Something about third leg jokes.
+			if(!(limb_type in OS.has_limbs))
+				var/obj/item/organ/external/E = H.get_organ(limb_type)
+				if(!E)
+					replace_after["limbs"][limb_type] = TRUE
+				else
+					remove_after["limbs"] |= E
+					H.bodyparts -= E
+					H.bodyparts_by_name -= E.limb_name
+					E.become_orphan()
+
+		for(var/B in H.bodyparts) //Is it mine?
+			var/obj/item/organ/external/E = B
+			if(!(E.limb_name in OS.has_limbs) && (E.type != OS.has_limbs[E.limb_name]["path"]))
+				continue
+			if(E.is_robotic() && (!ismachine(H) || isrobot(initial(E)))) //FBPs are fair game.
```

Comment:
Woops. I'll remove the snowflake check, especially since I added `rebuild_on_gain` to IPCs.